engine/builtin.go

```diff
@@ -813,6 +813,50 @@ func Compare(order, term1, term2 Term, k func(*Env) *Promise, env *Env) *Promise
 	}
 }
 
+// Between succeeds when lower, upper, and value are all integers, and lower <= value <= upper.
+// If value is a variable, it is unified with successive integers from lower to upper.
+func Between(lower, upper, value Term, k func(*Env) *Promise, env *Env) *Promise {
+	var low, high Integer
+
+	switch lower := env.Resolve(lower).(type) {
+	case Integer:
+		low = lower
+	case Variable:
+		return Error(ErrInstantiation)
+	default:
+		return Error(TypeErrorInteger(lower))
+	}
+
+	switch upper := env.Resolve(upper).(type) {
+	case Integer:
+		high = upper
+	case Variable:
+		return Error(ErrInstantiation)
+	default:
+		return Error(TypeErrorInteger(upper))
+	}
+
+	if low > high {
+		return Bool(false)
+	}
+
+	switch value := env.Resolve(value).(type) {
+	case Integer:
+		if value >= low && value <= high {
```

Comment:
This is a minor point but it'd be better to flow to `k(env)`.

> The code reads well if the successful flow of control runs down the page, eliminating error cases as they arise.
https://go.dev/doc/effective_go#if

```
  *
  |
  +--> Bool(false)
  |
  +--> Error(err)
  |
  v
k(env)
```

Reply:
Good point! I tweaked it a bit. Should we go further and put k(env) at the very end? I'm OK with either way.